guides/source/action_controller_advanced_topics.md

```diff
@@ -0,0 +1,500 @@
+
+    song.each do |line|
+      response.stream.write line.lyrics
+      sleep line.num_beats
+    end
+  ensure
+    response.stream.close
+  end
+end
+```
+
+#### Streaming Considerations
+
+Streaming arbitrary data is an extremely powerful tool. As shown in the previous
+examples, you can choose when and what to send across a response stream.
+However, you should also note the following things:
+
+* Each response stream creates a new thread and copies over the thread local
+  variables from the original thread. Having too many thread local variables can
+  negatively impact performance. Similarly, a large number of threads can also
+  hinder performance.
+* Failing to close the response stream will leave the corresponding socket open
+  forever. Make sure to call `close` whenever you are using a response stream.
+* WEBrick servers buffer all responses, and so streaming with
+  `ActionController::Live` will not work. You must use a web server which does
+  not automatically buffer responses.
+
+Log Filtering
+-------------
+
+Rails keeps a log file for each environment in the `log` folder at the application's root directory. Log files are extremely useful when debugging your application, but in a production environment you may not want every bit of information stored in log files. Rails allows you to specify parameters that should not be stored.
+
+### Parameters Filtering
+
+You can filter out sensitive request parameters from your log files by
+appending them to [`config.filter_parameters`][] in the application configuration.
+
+```ruby
+config.filter_parameters << :password
+```
+
+These parameters will be marked `[FILTERED]` in the log.
+
+The parameters specified in `filter_parameters` will be filtered out with
+partial matching regular expression. So for example, `:passw` will filter out `password`, `password_confirmation`, etc.
+
+Rails adds a list of default filters, including `:passw`, `:secret`, and
+`:token`, in the appropriate initializer
+(`initializers/filter_parameter_logging.rb`) to handle typical application
+parameters like `password`, `password_confirmation` and `my_token`.
+
+[`config.filter_parameters`]: configuring.html#config-filter-parameters
+
+### Redirects Filtering
+
+Sometimes it's desirable to filter out sensitive locations that your application is redirecting to. You can do that by using the `config.filter_redirect` configuration option:
+
+```ruby
+config.filter_redirect << "s3.amazonaws.com"
+```
+
+You can set it to a String, a Regexp, or an array of both.
+
+```ruby
+config.filter_redirect.concat ["s3.amazonaws.com", /private_path/]
+```
+
+Matching URLs will be replaced with `[FILTERED]`. However, if you only wish to filter the parameters, not the whole URLs, you can use parameter filtering.
+
+Force HTTPS Protocol
+--------------------
+
+If you'd like to ensure that communication to your controller is only possible
+via HTTPS, you can do so by enabling the [`ActionDispatch::SSL`][] middleware
```

Comment:
I think technically it just forces TLS, so protocols like WSS are allowed in addition to HTTPS. Maybe not worth noting.

Reply:
Agree, though not sure it's worth noting.